File: PluginMessagingUtils/src/main/java/org/openredstone/messages/ActionMessage.java
package org.openredstone.messages;

import org.openredstone.types.Action;

import java.util.Arrays;
import java.util.UUID;

public class ActionMessage extends Message {

    Action action;

    public ActionMessage(Action action, UUID uuid, String[] arguments) {
        super(uuid, arguments);
        this.action = action;
    }

    public ActionMessage(String serializedMessage) throws Exception {
        String[] raw = serializedMessage.split(":");

        if (raw.length < 2) {
            throw new Exception("Not enough arguments provided in serialized message.");
        }

        this.action = Action.valueOf(raw[0]);
        this.uuid = UUID.fromString(raw[1]);

        if (!isValidAction(raw[0])) {
            throw new IllegalArgumentException("Invalid action: " + raw[0]);
        }

        if(!isValidUuid()) {
            throw new IllegalArgumentException("Invalid uuid: " + raw[1]);
        }
        this.arguments = Arrays.copyOfRange(raw, 2, raw.length);

    }

    @Override
    public String getSerializedMessage() {
        String actionString = action.name();
        return actionString + ":" + super.getSerializedMessage();
    }

    private static boolean isValidAction(String action) {
        try {
            Action.valueOf(action);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    public Action getAction() {
        return action;
    }

}

File: plugin-messaging/src/main/kotlin/org/openredstone/messages/ActionMessage.kt
package org.openredstone.messages

import org.openredstone.types.Action
import java.util.*

class ActionMessage {
    var arguments: Array<String>
    var action: Action
    var uuid: UUID

    constructor(action: Action, uuid: UUID, arguments: Array<String>) {
        this.arguments = arguments
        this.action = action
        this.uuid = uuid
    }

    constructor(serializedMessage: String) {
        val raw: Array<String> = serializedMessage.split(":").toTypedArray()
        if (raw.size < 2) {
            throw Exception("Not enough arguments provided in serialized message.")
        }
        action = Action.valueOf(raw[0])
        uuid = UUID.fromString(raw[1])
        require(isValidAction(raw[0])) { "Invalid action: " + raw[0] }
        require(isValidUuid) { "Invalid uuid: " + raw[1] }
        arguments = raw.copyOfRange(2, raw.size)
    }

    fun getSerializedMessage(): String =
        action.name + ":" + uuid.toString() + ":" + arguments.joinToString(":")


    private val isValidUuid: Boolean
        get() = uuid.toString().matches("[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}".toRegex())

    companion object {
        private fun isValidAction(action: String?): Boolean {
            return try {
                Action.valueOf(action!!)
                true
            } catch (e: IllegalArgumentException) {
                false
            }
        }
    }
}
